store/tikv/lock_resolver.go

```diff
@@ -578,10 +604,266 @@ func (lr *LockResolver) getTxnStatus(bo *Backoffer, txnID uint64, primary []byte
+	// secondaries.
+	if primaryKey != nil {
+		keys = append(keys, primaryKey)
+	}
+
+	data.keys[curRegionID] = keys
+	return nil
+}
+
+func (lr *LockResolver) checkSecondaries(bo *Backoffer, txnID uint64, curKeys [][]byte, primaryKey []byte, curRegionID RegionVerID, shared *asyncResolveData) error {
+	checkReq := &kvrpcpb.CheckSecondaryLocksRequest{
+		Keys:         curKeys,
+		StartVersion: txnID,
+	}
+	req := tikvrpc.NewRequest(tikvrpc.CmdCheckSecondaryLocks, checkReq)
+	tikvLockResolverCountWithQueryCheckSecondaryLocks.Inc()
+	resp, err := lr.store.SendReq(bo, req, curRegionID, readTimeoutShort)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	regionErr, err := resp.GetRegionError()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if regionErr != nil {
+		err = bo.Backoff(BoRegionMiss, errors.New(regionErr.String()))
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		logutil.BgLogger().Info("checkSecondaries region error, regrouping", zap.Uint64("txn id", txnID), zap.Uint64("region", curRegionID.GetID()))
+
+		// If regions have changed, then we might need to regroup the keys. Since this should be rare and for the sake
+		// of simplicity, we will resolve regions sequentially.
+		regions, _, err := lr.store.GetRegionCache().GroupKeysByRegion(bo, curKeys, nil)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		for regionID, keys := range regions {
+			// Recursion will terminate because the resolve request succeeds or the Backoffer reaches its limit.
+			if err = lr.checkSecondaries(bo, txnID, keys, primaryKey, regionID, shared); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+	if resp.Resp == nil {
+		return errors.Trace(ErrBodyMissing)
+	}
+
+	checkResp := resp.Resp.(*kvrpcpb.CheckSecondaryLocksResponse)
+	return shared.addKeys(checkResp.Locks, curRegionID, len(curKeys), primaryKey, txnID, checkResp.CommitTs)
+}
+
+// resolveLockAsync resolves l assuming it was locked using the async commit protocol.
+func (lr *LockResolver) resolveLockAsync(bo *Backoffer, l *Lock, status TxnStatus) error {
+	tikvLockResolverCountWithResolveAsync.Inc()
+
+	regions, _, err := lr.store.GetRegionCache().GroupKeysByRegion(bo, status.secondaries, nil)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// We treat the primary key a bit differently to potentially avoid sending an unnecessary CheckSecondaryLocks message.
+	primaryLoc, err := lr.store.GetRegionCache().LocateKey(bo, l.Primary)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	shared := asyncResolveData{
+		mutex:       sync.Mutex{},
+		commitTs:    status.primaryLock.MinCommitTs,
+		keys:        map[RegionVerID][][]byte{},
+		missingLock: false,
+	}
+	var foundPrimaryRegion bool
+
+	errChan := make(chan error, len(regions))
+
+	for regionID, keys := range regions {
+		var primaryKey []byte
+		if regionID == primaryLoc.Region {
+			primaryKey = l.Primary
+			foundPrimaryRegion = true
+		}
```

Comment:
I don't understand it. What's the difference if we directly add the primary lock to `shared.keys[primaryLoc.Region]`?